apollo-router/src/plugins/cache/entity.rs

```diff
@@ -0,0 +1,618 @@
+    }
+
+    Ok(())
+}
+
+pub(crate) fn hash_vary_headers(headers: &http::HeaderMap) -> String {
+    let mut digest = Sha256::new();
+
+    for vary_header_value in headers.get_all(header::VARY).into_iter() {
+        if vary_header_value == "*" {
+            return String::from("*");
+        } else {
+            let header_names = match vary_header_value.to_str() {
+                Ok(header_val) => header_val.split(", "),
+                Err(_) => continue,
+            };
+            header_names.for_each(|header_name| {
+                if let Some(header_value) = headers.get(header_name).and_then(|h| h.to_str().ok()) {
+                    digest.update(header_value);
+                    digest.update(&[0u8; 1][..]);
+                }
+            });
+        }
+    }
+
+    hex::encode(digest.finalize().as_slice())
+}
+
+pub(crate) fn hash_query(body: &graphql::Request) -> String {
+    let mut digest = Sha256::new();
+    digest.update(body.query.as_deref().unwrap_or("-").as_bytes());
+    digest.update(&[0u8; 1][..]);
+    digest.update(body.operation_name.as_deref().unwrap_or("-").as_bytes());
+    digest.update(&[0u8; 1][..]);
+
+    hex::encode(digest.finalize().as_slice())
+}
+
+pub(crate) fn hash_additional_data(body: &mut graphql::Request, context: &Context) -> String {
+    let mut digest = Sha256::new();
+
+    let repr_key = ByteString::from(REPRESENTATIONS);
+    // Removing the representations variable because it's already part of the cache key
+    let representations = body.variables.remove(&repr_key);
+    digest.update(&serde_json::to_vec(&body.variables).unwrap());
+    if let Some(representations) = representations {
+        body.variables.insert(repr_key, representations);
+    }
+
+    let cache_key = context
+        .private_entries
+        .lock()
+        .get::<CacheKeyMetadata>()
+        .cloned()
+        .unwrap_or_default();
+    digest.update(&serde_json::to_vec(&cache_key).unwrap());
+
+    hex::encode(digest.finalize().as_slice())
+}
+
+// build a cache key for the root operation
+fn extract_cache_key_root(
+    subgraph_name: &str,
+    body: &mut graphql::Request,
+    context: &Context,
+) -> String {
+    // hash the query and operation name
+    let query_hash = hash_query(body);
+    // hash more data like variables and authorization status
+    let additional_data_hash = hash_additional_data(body, context);
+
+    // the cache key is written to easily find keys matching a prefix for deletion:
+    // - subgraph name: caching is done per subgraph
+    // - query hash: invalidate the entry for a specific query and operation name
+    // - additional data: separate cache entries depending on info like authorization status
+    format!(
+        "subgraph.{}|{}|{}",
+        subgraph_name, query_hash, additional_data_hash
+    )
+}
+
+// build a list of keys to get from the cache in one query
+fn extract_cache_keys(
+    //representations: &mut Vec<Value>,
```

Comment:
To delete

Reply:
fixed in [b529853](https://github.com/apollographql/router/pull/4170/commits/b5298535ac0b1754e975ae139831f462d467fe5b)